modules/aui2s/aui2s_play.c

```diff
@@ -0,0 +1,157 @@
+/**
+ * @file aui2s_play.c freeRTOS I2S audio driver module - player
+ *
+ * Copyright (C) 2019 cspiel.at
+ */
+#include <sys/types.h>
+#include <sys/time.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <pthread.h>
+#include "freertos/FreeRTOS.h"
+#include "driver/i2s.h"
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include "aui2s.h"
+
+
+struct auplay_st {
+	const struct auplay *ap;  /* pointer to base-class (inheritance) */
+	pthread_t thread;
+	bool run;
+	void *sampv;
+	size_t sampc;
+	size_t bytes;
+	auplay_write_h *wh;
+	void *arg;
+	struct auplay_prm prm;
+
+	uint32_t *pcm;
+};
+
+
+static void auplay_destructor(void *arg)
+{
+	struct auplay_st *st = arg;
+
+	/* Wait for termination of other thread */
+	if (st->run) {
+		info("aui2s: stopping playback thread\n");
+		st->run = false;
+		(void)pthread_join(st->thread, NULL);
+	}
+
+	mem_deref(st->sampv);
+	mem_deref(st->pcm);
+}
+
+
+static void convert_sampv(struct auplay_st *st, size_t i, size_t n)
+{
+	uint32_t j;
+	int16_t *sampv = st->sampv;
+	for (j = 0; j < n; j++) {
+		uint32_t v = sampv[i+j];
+		st->pcm[j] = v << 17;
+	}
+}
+
+static void *write_thread(void *arg)
+{
+	struct auplay_st *st = arg;
+	int err;
+
+	err = aui2s_start(st->prm.srate, I2O_PLAY);
+	if (err) {
+		warning("aui2s: could not start auplay\n");
+		return NULL;
+	}
+
+	i2s_set_clk(I2S_PORT, st->prm.srate, 32, 1);
```

Comment:
the last parameter (1) is number of channels. perhaps make a constant for that,
and also check number of channels in the alloc functions.